libs/ballot-interpreter-nh-next/src/metadata.rs

```diff
@@ -0,0 +1,354 @@
+}
+
+impl Debug for BallotPageMetadataBack {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("BackMetadata")
+            .field("bits", &print_boolean_slice_as_binary(&self.bits))
+            .field("election_day", &self.election_day)
+            .field("election_month", &self.election_month)
+            .field("election_year", &self.election_year)
+            .field("election_type", &self.election_type)
+            .field(
+                "ender_code",
+                &print_boolean_slice_as_binary(&self.ender_code),
+            )
+            .field(
+                "expected_ender_code",
+                &print_boolean_slice_as_binary(&self.expected_ender_code),
+            )
+            .finish()
+    }
+}
+
+#[derive(Debug, Serialize)]
+#[serde(tag = "side", rename_all = "camelCase")]
+pub enum BallotPageMetadata {
+    Front(BallotPageMetadataFront),
+    Back(BallotPageMetadataBack),
+}
+
+#[derive(Debug, Serialize)]
+#[serde(tag = "type", rename_all = "camelCase")]
+pub enum BallotPageMetadataError {
+    ValueOutOfRange {
+        field: String,
+        value: u32,
+        min: u32,
+        max: u32,
+        metadata: BallotPageMetadata,
+    },
+    InvalidChecksum {
+        metadata: BallotPageMetadataFront,
+    },
+    InvalidEnderCode {
+        metadata: BallotPageMetadataBack,
+    },
+    InvalidTimingMarkCount {
+        expected: usize,
+        actual: usize,
+    },
+    AmbiguousMetadata {
+        front_metadata: BallotPageMetadataFront,
+        back_metadata: BallotPageMetadataBack,
+    },
+}
+
+/// Computes the metadata bits from the bottom row of a ballot page.
+pub fn compute_bits_from_bottom_timing_marks(
+    geometry: &Geometry,
+    bottom_timing_marks: &[Option<Rect>],
+) -> Result<[bool; METADATA_BITS], BallotPageMetadataError> {
+    // Validate the number of timing marks.
+    if bottom_timing_marks.len() != geometry.grid_size.width as usize
+        || geometry.grid_size.width as usize != METADATA_BORDER_MARK_COUNT
+    {
+        return Err(BallotPageMetadataError::InvalidTimingMarkCount {
+            expected: geometry.grid_size.width as usize,
+            actual: bottom_timing_marks.len(),
+        });
+    }
+
+    // The first and last timing marks must be present.
+    match (bottom_timing_marks.first(), bottom_timing_marks.last()) {
+        (Some(Some(_)), Some(Some(_))) => {}
+        _ => {
+            return Err(BallotPageMetadataError::InvalidTimingMarkCount {
+                expected: 2,
+                actual: bottom_timing_marks
+                    .iter()
+                    .filter(|&mark| mark.is_some())
+                    .count(),
+            });
+        }
+    }
+
+    let bits: Vec<bool> = bottom_timing_marks
+        .iter()
+        .skip(1)
+        .take(METADATA_BITS)
+        .map(|timing_mark| timing_mark.is_some())
+        .rev()
+        .collect();
+
+    let bit_count = bits.len();
+    bits.try_into()
```

Comment:
Took me a while to figure out what this was doing (checking the length of the bits in a type-safe way). Maybe a comment could be helpful?

Reply:
Added a comment.